src/normalized_cut.jl

```diff
@@ -0,0 +1,132 @@
+"""
+```
+labels                = ncut(W, [num_cuts], [thres])
+```
+
+Segments graph represented by adjacency matrix W using recursive two-way normalized cut. Returns segment label for each node.
+
+Parameters:  
+-    W              = Weighted (sparse) adjacency matrix for graph.
+-    thres          = Subgraphs aren't split if best normalized cut is above this threshold
+-    num_cuts       = Number of cuts performed to determine optimal cut
+
+# Example
+```julia
+using LightGraphs, SimpleWeightedGraphs, ImageSegmentation
+g = SimpleWeightedGraph(6)
+add_edge!(g, 1, 2, 1)
+add_edge!(g, 2, 3, 1)
+add_edge!(g, 1, 3, 1)
+add_edge!(g, 4, 5, 1)
+add_edge!(g, 4, 6, 1)
+add_edge!(g, 5, 6, 1)
+add_edge!(g, 1, 6, 0.1)
+add_edge!(g, 3, 4, 0.2)
+
+labels = ncut(g.weights, 0.1)
+```
+"""
+function ncut{T<:Real}(W::SparseMatrixCSC{T,Int}, thres::Real = 0.01 ,num_cuts::Int = 10)
+
+    #returns normalized cut score
+    function ncut_cost(mask)
+        cut_cost = 0
+        rows = rowvals(W)
+        vals = nonzeros(W)
+        for i = 1:n
+            for j in nzrange(W, i)
+                row = rows[j]
+                if mask[i] != mask[row]
+                    cut_cost += vals[j]/2
+                end
+            end
+        end
+        return cut_cost/sum(D*mask) + cut_cost/sum(D*(.~mask))
+    end
+
+    m, n = size(W)
+    D = Diagonal(vec(sum(W, 2)))
+
+    if m == 1
+        return [1]
+    end
+
+    #get eigenvector corresponding to second smallest eigenvalue
+    v = eigvecs(full(D-W), full(D))[:, 2]
+
+    #perform n-cuts with different partitions of v and find best one
+    min_cost = Inf
+    best_thres = -1
+    for t in linspace(minimum(v), maximum(v), num_cuts)
+        mask = v.>t
+        cost = ncut_cost(mask)
+        if cost < min_cost
+            min_cost = cost
+            best_thres = t
+        end
+    end
```

Comment:
You can optimize this to use matrix permutation and one matrix vector product to calculate all `nv(g)` possible cuts at once. But this one works well enough for large `ncuts`.